### PR TITLE
New weave implementation and update for dc.molnet

### PR DESCRIPTION
**Pull request by miaecle**
This PR adds in:

- Save and reload featurized datasets to(from) DEEPCHEM_DATA_DIR

- A new implementation for weave model(weave_v2), approximately 2 to 6 times faster

- xgboost, DAG, DTNN, weave(v2) models in dc.molnet module

- performances for new models

Dataset saving and reloading(#494) may take quite a amount of disk space, can be switched off by setting `reload=False` and deleting the files.
@rbharath I did not find a internal way to save `transformers`, hence use pickle instead, any idea about a better method to do this?

### CODE REVIEW

File: deepchem/molnet/load_function/bace_datasets.py [hunk 1]
```diff
@@ -80,6 +108,21 @@ def load_bace_classification(featurizer=None, split='random'):
     )
 
   bace_tasks = ["Class"]
+  if save:
```
Comment (rbharath) [inline deepchem/molnet/load_function/bace_datasets.py]: These save functions are pretty similar for different datasets. Might be good to factor them out into a `dc.molnets.utils` file with a `save_dataset()` function that can be reused.

**Comment by rbharath** [comment]
@miaecle Thanks for getting this started! I think saving the transformers with Pickles should be fine for now.

As a general request, would you mind adding a unit test for the saving/reload capabilities? Perhaps with a dummy `dc.molnet` dataset with 10 datapoints or the like.

File: deepchem/molnet/load_function/bace_datasets.py (hunk 1 above)
Comment (miaecle) [inline deepchem/molnet/load_function/bace_datasets.py]: Good point, will do!

**Comment by XericZephyr** [comment]
hi, @miaecle ,

you might consider my suggestion about `joblib.Memory`. The idea is to cache the result according to the hashable input, e.g., split, featurizer, and so.

You can do cache in the loading function dict, e.g., 

```python
mem = joblib.Memory(cachedir="/tmp/deepchem/")

cached_loading_functions = {k: mem.cache(v) for k, v in loading_functions}
``` 

This modification **eases your pain** to maintain so many loading function files.

**Comment by miaecle** [comment]
@rbharath Could you take a look? I have basically finished on this PR. There is also a major update on weave models, I have added another implementation which is significantly faster(now 1.5~2 times slower than pure graphconv). Performances look pretty decent, especially on scaffold splitting.

File: deepchem/nn/weave_layers.py [hunk 2]
```diff
@@ -154,21 +158,87 @@ def call(self, x, mask=None):
...
+      P = tf.tensordot(tf.concat([AP, PP], 3), self.W_P, [[3], [0]]) + self.b_P
+      P = self.activation(P)
+      P = tf.multiply(P, tf.expand_dims(pair_mask, axis=3))
+    else:
+      P = pair_features
+
+    return A, P
+
+
+class WeaveLayer_v2(WeaveLayer):
```
Comment (rbharath) [inline deepchem/nn/weave_layers.py]: Same comment. Perhaps `AlternateWeaveLayer`?

**Comment by rbharath** [comment]
@miaecle Thanks for quick turnaround! Should be good to merge once tests pass.

**Comment by miaecle** [comment]
@rbharath Sounds good, thanks for reviewing the code!

**Pull request merged**